hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemChooseSAS.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
+import org.junit.Assume;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.apache.hadoop.fs.azurebfs.constants.ConfigurationKeys.FS_AZURE_SAS_FIXED_TOKEN;
+import static org.apache.hadoop.fs.azurebfs.constants.ConfigurationKeys.FS_AZURE_SAS_TOKEN_PROVIDER_TYPE;
+import static org.apache.hadoop.test.LambdaTestUtils.intercept;
+
+public class ITestAzureBlobFileSystemChooseSAS extends AbstractAbfsIntegrationTest{
+
+    private String accountSAS;
+
+    public ITestAzureBlobFileSystemChooseSAS() throws Exception {
+        // The test uses shared key to create a random filesystem and then creates another
+        // instance of this filesystem using SAS authorization.
+        Assume.assumeTrue(this.getAuthType() == AuthType.SharedKey);
+    }
+
+    private void generateAccountSAS() throws AzureBlobFileSystemException {
+        final String accountKey = getConfiguration().getStorageAccountKey();
+        AccountSASGenerator configAccountSASGenerator = new AccountSASGenerator(Base64.decode(accountKey));
+        accountSAS = configAccountSASGenerator.getAccountSAS(getAccountName());
+    }
+
+    @Override
+    public void setup() throws Exception {
+        createFilesystemForSASTests();
+        super.setup();
+        // obtaining an account SAS token from in-built generator to set as configuration for testing filesystem level operations
+        generateAccountSAS();
+    }
+
+    /**
+     * Tests the scenario where both the token provider class and a fixed token are configured:
+     * whether the correct choice is made (precedence given to token provider class), and the chosen SAS Token works as expected
+     * @throws Exception
+     */
+    @Test
+    public void testBothProviderFixedTokenConfigured() throws Exception {
+        AbfsConfiguration testAbfsConfig = getConfiguration();
+
+        // configuring a SASTokenProvider class: this provides a user delegation SAS
+        // user delegation SAS Provider is set
+        // This easily distinguishes between results of filesystem level and blob level operations to ensure correct SAS is chosen,
+        // when both a provider class and fixed token is configured.
+        testAbfsConfig.set(FS_AZURE_SAS_TOKEN_PROVIDER_TYPE, "org.apache.hadoop.fs.azurebfs.extensions.MockDelegationSASTokenProvider");
+
+        // configuring the fixed SAS token
+        testAbfsConfig.set(FS_AZURE_SAS_FIXED_TOKEN, accountSAS);
+
+        // creating a new fs instance with the updated configs
+        AzureBlobFileSystem newTestFs = (AzureBlobFileSystem) FileSystem.newInstance(testAbfsConfig.getRawConfiguration());
+
+        // testing a file system level operation
+        TracingContext tracingContext = getTestTracingContext(newTestFs, true);
+        // expected to fail in the ideal case, as delegation SAS will be chosen, provider class is given preference when both are configured
```

Comment:
why is it expected to fail. Agree DelegationSAS will be taken, but what is the point which would raise exception.

Reply:
This is expected to fail because the filesystem level operations are beyond the scope of a Delegation SAS Token. Added a comment for this.